subprojects/dependency-management/src/integTest/groovy/org/gradle/integtests/resolve/transform/ArtifactTransformBuildOperationIntegrationTest.groovy

```diff
@@ -0,0 +1,609 @@
+        setupExternalDependency()
+
+        buildFile << """
+            allprojects {
+                dependencies {
+                    registerTransform(MakeGreen) {
+                        from.attribute(color, 'blue')
+                        to.attribute(color, 'green')
+                    }
+                }
+            }
+
+            project(":consumer") {
+                dependencies {
+                    implementation project(":producer")
+                }
+            }
+        """
+
+        when:
+        run ":consumer:resolve"
+
+        then:
+        executedAndNotSkipped(":consumer:resolve")
+
+        result.groupedOutput.transform("MakeGreen")
+            .assertOutputContains("processing producer.jar using [test-4.2.jar]")
+
+        result.groupedOutput.task(":consumer:resolve")
+            .assertOutputContains("result = [producer.jar.green, test-4.2.jar]")
+
+        def plannedNodes = buildOperations.only(CalculateTaskGraphBuildOperationType).result.executionPlan as List<PlannedNode>
+        plannedNodes.every { KNOWN_NODE_TYPES.contains(it.nodeIdentity.nodeType) }
+        plannedNodes.count { it.nodeIdentity.nodeType.toString() == ARTIFACT_TRANSFORM } == 1
+
+        checkExecutionPlanMatchingDependencies(
+            plannedNodes,
+            [
+                taskMatcher(
+                    "node1", ":producer:producer",
+                    []
+                ),
+                transformMatcher(
+                    "node2", ":consumer", [buildPath: ":", projectPath: ":producer"], [artifactType: "jar", color: "green"],
+                    ["node1"]
+                ),
+                taskMatcher(
+                    "node3", ":consumer:resolve",
+                    ["node2"]
+                ),
+            ]
+        )
+
+        with(buildOperations.only(ExecuteScheduledTransformationStepBuildOperationType).details) {
+            transformerName == "MakeGreen"
+            subjectName == "producer.jar (project :producer)"
+            with(transformationIdentity) {
+                nodeType == "ARTIFACT_TRANSFORM"
+                buildPath == ":"
+                projectPath == ":consumer"
+                targetVariant.componentId == [buildPath: ":", projectPath: ":producer"]
+                targetVariant.attributes == [color: "green", artifactType: "jar"]
+                targetVariant.capabilities == [[group: "colored", name: "producer", version: "unspecified"]]
+                artifactName == "producer.jar"
+                dependenciesConfigurationIdentity == [buildPath: ":", projectPath: ":consumer", name: "resolver"]
+            }
+            transformType == "MakeGreen"
+            sourceAttributes == [color: "blue", artifactType: "jar"]
+            fromAttributes == [color: "blue"]
+            toAttributes == [color: "green"]
+        }
+    }
+
+    def "chained transform with upstream dependencies operations are captured"() {
```

Comment:
❌ I suppose we should also have a test where the initial component has two artifacts.